server/channels/store/sqlstore/draft_store.go

```diff
@@ -236,3 +233,72 @@ func (s *SqlDraftStore) determineMaxDraftSize() int {
 
 	return maxDraftSize
 }
+
+func (s *SqlDraftStore) GetLastCreateAtAndUserIdValuesForEmptyDraftsMigration(createAt int64, userId string) (int64, string, error) {
+	var drafts []struct {
+		CreateAt int64
+		UserId   string
+	}
+
+	query := s.getQueryBuilder().
+		Select("CreateAt", "UserId").
+		From("Drafts").
+		Where(sq.Or{
+			sq.Gt{"CreateAt": createAt},
+			sq.And{
+				sq.Eq{"CreateAt": createAt},
+				sq.Gt{"UserId": userId},
+			},
+		}).
+		OrderBy("CreateAt", "UserId ASC").
+		Limit(100)
+
+	err := s.GetReplicaX().SelectBuilder(&drafts, query)
+	if err != nil {
+		return 0, "", errors.Wrap(err, "failed to get the list of drafts")
+	}
+
+	if len(drafts) == 0 {
+		return 0, "", nil
+	}
+
+	lastElement := drafts[len(drafts)-1]
+	return lastElement.CreateAt, lastElement.UserId, nil
+}
+
+func (s *SqlDraftStore) DeleteEmptyDraftsByCreateAtAndUserId(createAt int64, userId string) error {
+
+	var sql string
+	if s.DriverName() == model.DatabaseDriverPostgres {
+		sql = `
+			WITH dd AS (
+				SELECT UserId, ChannelId, RootId FROM Drafts 
+				WHERE (CreateAt > ? OR (CreateAt = ? AND UserId > ?)) 
+				ORDER BY CreateAt, UserId ASC LIMIT 100
+			) 
+			DELETE FROM
+				Drafts d 
+			USING dd 
+			WHERE 
+				d.UserId = dd.UserId 
+				AND d.ChannelId = dd.ChannelId 
+				AND d.RootId = dd.RootId
+				AND d.Message = ''
```

Comment:
The problem is to store the offsets. You could do everything in one query, but then MySQL does not support `RETURNING` in the DELETE query. So you need one query to calculate the offsets, another to DELETE.

Reply:
Do we care about the offsets? Is that just used to report job metadata?

Reply:
We care about offsets to survive server restarts. It is an important aspect to make the job resumable. You can restart the server in the middle of a migration, and it picks right off from where it left.

Reply:
Apologies that I'm still missing something, @agnivade. Here's a more concrete pseudo-code example of my suggestion:

```go
for {
    numRowsDeleted, err := sql.Execute("DELETE FROM Drafts d WHERE d.Message = '' LIMIT 100")
    if numRowsDeleted == 0 {
        break
    }

    time.Sleep(1 * time.Second)
}
```

This pseudo-code will eventually delete all messages with empty drafts in chunks of 100 at a time. It doesn't care about offsets, because it doesn't care about which messages it needs to delete: just that they all go away. It doesn't matter if the code gets restarted: if there are any messages that still need to be deleted, it will find them and pick up from there.

One downside to this approach is presumably the need to scan the table for `d.Message = ''` in the first place. But given we're deleting the rows behind us, and we already know there are a disproportionate number of empty messages compared to "real drafts", we need to look at most of the rows anyway.

Thoughts?

Reply:
I'm wondering if the offsets are to avoid locking the whole table every time while the job is running? Not sure if the table is partitioned or not (could be under the hood) but the offsets make sure that we're only working with a specific data set, so the DB won't lock the whole table.

It also means that we can leverage the indexes so that we can grab the rows we want to work with quickly, then we only scan 100 rows at a time maximum while checking for blanks, whereas we'd be scanning more without the offsets.

Reply:
A simpler reason for Postgres:
```
ERROR:  syntax error at or near "LIMIT"
LINE 1: DELETE FROM Drafts d where d.Message = '' LIMIT 100
```

I'm guessing that batching is the only way for Postgres?

Reply:
@lieut-data - `LIMIT` on DELETEs is only supported for MySQL unfortunately. :(

But even if we hadn't supported Postgres, my main objective here was to create an example migration that can serve as a template for future ones. I think overall, this is a pretty simple and straightforward migration which shouldn't take too long.